Rewrite example source to make it simpler and clearer. (#33339)
dotnet/docs: docs/core/extensions/dependency-injection-usage.md

Using either the [dotnet new](../tools/dotnet-new.md) command or an IDE new project wizard, create a new .NET console application named **ConsoleDI.Example**. Add the [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) NuGet package to the project.

Your new console app project file should resemble the following:

:::code language="xml" source="snippets/configuration/console-di/console-di.csproj":::

> [!IMPORTANT]

> > In this example, the [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) NuGet package is required to build and run the app. Some metapackages might contain the `Microsoft.Extensions.Hosting` package, in those scenarios and explicit package reference isn't required.

Reviewer: ```suggestion
> In this example, the [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) NuGet package is required to build and run the app. Some metapackages might contain the `Microsoft.Extensions.Hosting` package, in which case an explicit package reference isn't required.
```

## Add interfaces

In this sample app, you'll learn how dependency injection handles service lifetime. You'll create several interfaces that represent different service lifetimes. Add the following interfaces to the project root directory:

*IReportServiceLifetime.cs*

:::code source="snippets/configuration/console-di/IReportServiceLifetime.cs":::

The `IReportServiceLifetime` interface defines:

- A `Guid Id` property that represents the unique identifier of the service.

> - The <xref:Microsoft.Extensions.DependencyInjection.ServiceLifetime> property that represents the service lifetime.

Reviewer: ```suggestion
- A <xref:Microsoft.Extensions.DependencyInjection.ServiceLifetime> property that represents the service lifetime.
```

## Add default implementations

The example implementations all initialize their `Id` property with the result of <xref:System.Guid.NewGuid?displayProperty=nameWithType>. Add the following default implementation classes for the various services to the project root directory:

*ExampleTransientService.cs*

:::code source="snippets/configuration/console-di/ExampleTransientService.cs":::

*ExampleScopedService.cs*

:::code source="snippets/configuration/console-di/ExampleScopedService.cs":::

*ExampleSingletonService.cs*

:::code source="snippets/configuration/console-di/ExampleSingletonService.cs":::

Each implementation is defined as `internal sealed` and implements its corresponding interface. For example, `ExampleSingletonService` implements `IExampleSingletonService`.

## Add a service that requires DI

Add the following service lifetime reporter class, which acts as a service to the console app:

*ServiceLifetimeReporter.cs*

:::code source="snippets/configuration/console-di/ServiceLifetimeReporter.cs":::

> The `ServiceLifetimeReporter` defines a constructor that requires each of the aforementioned marker interfaces, that is, `IExampleTransientService`, `IExampleScopedService`, and `IExampleSingletonService`. The object exposes a single method that allows the consumer to report on the service with a given `lifetimeDetails` parameter. When invoked, the `ReportServiceLifetimeDetails` method logs each service's unique identifier with the service lifetime message.

Reviewer: It looks like you removed the line (65) where marker interfaces were previously mentioned, so I think some further explanation is needed.

> The `ServiceLifetimeReporter` defines a constructor that requires each of the aforementioned marker interfaces, that is, `IExampleTransientService`, `IExampleScopedService`, and `IExampleSingletonService`. The object exposes a single method that allows the consumer to report on the service with a given `lifetimeDetails` parameter. When invoked, the `ReportServiceLifetimeDetails` method logs each service's unique identifier with the service lifetime message.

Author: ```suggestion
The `ServiceLifetimeReporter` defines a constructor that requires each of the aforementioned service interfaces, that is, `IExampleTransientService`, `IExampleScopedService`, and `IExampleSingletonService`. The object exposes a single method that allows the consumer to report on the service with a given `lifetimeDetails` parameter. When invoked, the `ReportServiceLifetimeDetails` method logs each service's unique identifier with the service lifetime message. The log messages help to visualize the service lifetime.
```

## Conclusion

> In this sample app, you created several interfaces and corresponding implementations. Each of these services is uniquely identified and paired with a <xref:Microsoft.Extensions.DependencyInjection.ServiceLifetime>. The sample app demonstrates registering service implementations against an interface, and how to register pure classes without backing interfaces. The sample app then demonstrates how dependencies defined as constructor parameters are resolved during runtime.

Reviewer: ```suggestion
In this sample app, you created several interfaces and corresponding implementations. Each of these services is uniquely identified and paired with a <xref:Microsoft.Extensions.DependencyInjection.ServiceLifetime>. The sample app demonstrates registering service implementations against an interface, and how to register pure classes without backing interfaces. The sample app then demonstrates how dependencies defined as constructor parameters are resolved at run time.
```

> When the app is ran, it will display output similar to the following:

Reviewer: ```suggestion
When you run the app, it displays output similar to the following:
```
